File: src/audioLeveling.ts
// Output-stage loudness leveling for the demodulated PCM.
//
// The demod modes leave the audio at wildly different levels: AM/CW are
// AGC'd inside the demodulator (to *different* setpoints, 16000 / 12000),
// while WFM / NFM / SSB are raw fixed-gain and ride signal strength + the
// RF-gain dial. Feeding those straight to the sink means band-to-band
// loudness jumps and deck-rx that's much quieter than other apps.
//
// Applied once just before audioOutput.write() (covers naudiodon + icecast):
//   1. per-mode makeup (MODE_MAKEUP) — a STATIC per-band gain that lifts each
//      mode to a common loudness. This is the PRIMARY leveller: it aligns the
//      bands and boosts the level with NO dynamic gain motion, so there's no
//      audible level-riding / pumping. Tunable per host via cfg.audioMakeup.
//   2. output AGC (OutputLeveler) — OPTIONAL (cfg.audioLeveling, default OFF).
//      A slow adaptive gain that also tracks within-band signal-strength
//      changes. Off by default: the dynamic level motion is audible and some
//      find it unpleasant, so the static makeup alone is the default.
//   3. soft limiter (softLimit) — instantaneous soft ceiling so the static
//      gain can run hot near full-scale without hard-clip distortion. Peak-
//      only; it does NOT cause the slow breathing an AGC does.
//
// cfg.audioGain is a master trim multiplied on top of the makeup (both modes).

/** Per-demod-mode STATIC makeup gain. Keyed on the numeric mode index
 *  (0=NFM 1=WFM 2=AM 3=DSB 4=USB 5=CW 6=LSB 7=RAW). These set the per-band
 *  output level directly (AGC off by default), so they're calibrated to bring
 *  each band to a common loudness. Starting estimates — AM/WFM are data-
 *  informed, the rest rougher; fine-tune by ear via cfg.audioMakeup (no
 *  rebuild) or here. CW is < 1 because the demod's CW AGC already normalises
 *  the BFO tone to ~12000. */
export const MODE_MAKEUP: Record<number, number> = {
  0: 5,    // NFM
  1: 10,   // WFM
  2: 3,    // AM
  3: 5,    // DSB (→ FM)
  4: 3,    // USB
  5: 0.6,  // CW (demod CW-AGC already at ~12000)
  6: 3,    // LSB
  7: 5,    // RAW (→ FM)
};

export const INT16_MAX = 32767;

/** Smooth soft-knee limiter. Linear below `ceil*kneeFrac`, then tanh-compressed
 *  up to (never past) `ceil`. C1-continuous at the knee (tanh'(0)=1 matches the
 *  linear slope) so there's no kink. Returns an integer in Int16 range. */
export function softLimit(x: number, ceil = INT16_MAX, kneeFrac = 0.85): number {
  const T = ceil * kneeFrac;
  const a = x < 0 ? -x : x;
  if (a <= T) return Math.round(x);
  const span = ceil - T;
  const compressed = T + span * Math.tanh((a - T) / span);
  return Math.round(x < 0 ? -compressed : compressed);
}

export interface LevelerCfg {
  /** Desired post-leveling RMS in Int16 units (~7500 ≈ -12.8 dBFS). */
  targetRms: number;
  /** Gain ceiling — caps how hard weak signals (and noise) get pushed. */
  maxGain: number;
  /** Gain floor — lets hot modes (SSB fixed 48000) be attenuated. */
  minGain: number;
  /** Time constant (s) for gain DECREASING (signal too loud) — fast. */
  attackTc: number;
  /** Time constant (s) for gain INCREASING (signal too quiet) — slow. */
  releaseTc: number;
  /** Below this makeup-applied input RMS, hold gain (don't pump up silence/noise). */
  noiseFloorRms: number;
  /** When false the AGC is bypassed (gain pinned to 1) — flat fallback. */
  enabled: boolean;
}

export const DEFAULT_LEVELER_CFG: LevelerCfg = {
  targetRms: 7500,
  maxGain: 16,
  minGain: 0.05,
  attackTc: 0.12,
  releaseTc: 1.5,
  noiseFloorRms: 40,
  enabled: true,
};

/** Stateful per-buffer output AGC. `gain` is the multiplier the caller applies
 *  to the makeup-scaled PCM. Drives buffer RMS toward cfg.targetRms with
 *  asymmetric attack/release; holds during near-silence and (caller's job)
 *  during mute. */
export class OutputLeveler {
  gain = 1;
  private cfg: LevelerCfg;

  constructor(cfg: Partial<LevelerCfg> = {}) {
    this.cfg = { ...DEFAULT_LEVELER_CFG, ...cfg };
  }

  configure(partial: Partial<LevelerCfg>): void {
    this.cfg = { ...this.cfg, ...partial };
  }

  getConfig(): Readonly<LevelerCfg> {
    return this.cfg;
  }

  /** RMS over all (interleaved) samples of the makeup-applied buffer. */
  static rms(pcm: Int16Array | number[], makeup: number): number {
    const n = pcm.length;
    if (n === 0) return 0;
    let sumSq = 0;
    for (let i = 0; i < n; i++) {
      const s = pcm[i] * makeup;
      sumSq += s * s;
    }
    return Math.sqrt(sumSq / n);
  }

  /** Advance `gain` one buffer toward the target. `dt` is the buffer duration
   *  in seconds (frames / audioRate). Returns the updated gain. */
  observe(pcm: Int16Array | number[], makeup: number, dt: number): number {
    if (!this.cfg.enabled) {
      this.gain = 1;
      return this.gain;
    }
    const rms = OutputLeveler.rms(pcm, makeup);
    if (rms < this.cfg.noiseFloorRms) return this.gain; // hold on near-silence
    let desired = this.cfg.targetRms / rms;
    if (desired > this.cfg.maxGain) desired = this.cfg.maxGain;
    else if (desired < this.cfg.minGain) desired = this.cfg.minGain;
    const tc = desired < this.gain ? this.cfg.attackTc : this.cfg.releaseTc;
    const alpha = dt > 0 && tc > 0 ? 1 - Math.exp(-dt / tc) : 1;
    this.gain += (desired - this.gain) * alpha;
    return this.gain;
  }

  /** Seed the gain straight to the target for this buffer (no time constant)
   *  — e.g. on unmute so audio returns at full level immediately instead of
   *  ramping up over the slow release TC. Holds on near-silence (nothing to
   *  seed against). Returns the updated gain. */
  snap(pcm: Int16Array | number[], makeup: number): number {
    if (!this.cfg.enabled) {
      this.gain = 1;
      return this.gain;
    }
    const rms = OutputLeveler.rms(pcm, makeup);
    if (rms < this.cfg.noiseFloorRms) return this.gain;
    let desired = this.cfg.targetRms / rms;
    if (desired > this.cfg.maxGain) desired = this.cfg.maxGain;
    else if (desired < this.cfg.minGain) desired = this.cfg.minGain;
    this.gain = desired;
    return this.gain;
  }

  reset(): void {
    this.gain = 1;
  }
}

File: native-app/Sources/AudioLeveling.swift
import Foundation

/// Output-stage loudness levelling, applied once just before the sink.
///
/// Port of `src/audioLeveling.ts`. Three stages, in this order and for these
/// reasons:
///
///   1. **Per-mode static makeup.** The demods leave audio at wildly different
///      levels — AM and CW are AGC'd inside the demodulator and to *different*
///      setpoints (16000 and 12000), while WFM, NFM and SSB are raw fixed gain
///      riding signal strength and the RF gain control. This is the primary
///      leveller: it aligns the bands with no dynamic motion at all, so there
///      is nothing to hear pumping.
///   2. **Adaptive output AGC — off by default.** It also tracks within-band
///      strength changes, but its level-riding is audible and not everyone
///      wants it. The static makeup alone is the default for that reason.
///   3. **Soft-knee limiter.** An instantaneous ceiling so the static gain can
///      run hot near full scale without hard-clipping. Peak-only, so it does
///      not breathe the way an AGC does.
enum AudioLeveling {

    /// Keyed by the SDR++ mode index — the numbering the plugin's MODE_MAKEUP
    /// uses, with the same values. Calibrated to bring each band to a common
    /// loudness with the AGC off. CW is below 1 because the demodulator's own
    /// AGC has already normalised the BFO tone.
    ///
    /// AM was 3 once: it over-drove envelope peaks into the limiter and the
    /// result was broadband distortion. 1.5 is about +3.5 dB and stays under
    /// the knee.
    static let modeMakeup: [Int: Double] = [
        0: 5,    // NFM
        1: 10,   // WFM
        2: 1.5,  // AM
        3: 5,    // DSB
        4: 3,    // USB
        5: 0.6,  // CW
        6: 3,    // LSB
        7: 5,    // RAW
    ]

    static let int16Max: Double = 32767

    /// Linear below `ceiling * kneeFrac`, then tanh-compressed up to — never
    /// past — the ceiling. C1-continuous at the knee, since tanh'(0) = 1
    /// matches the linear slope, so there is no kink to hear.
    static func softLimit(_ x: Double, ceiling: Double = int16Max,
                          kneeFrac: Double = 0.85) -> Double {
        let t = ceiling * kneeFrac
        let a = abs(x)
        if a <= t { return x }
        let span = ceiling - t
        let compressed = t + span * tanh((a - t) / span)
        return x < 0 ? -compressed : compressed
    }
}

/// Per-buffer output AGC. `gain` is what the caller multiplies onto the
/// makeup-scaled samples.
final class OutputLeveler {

    struct Config {
        /// Post-levelling RMS in int16 units. 7500 is about -12.8 dBFS.
        var targetRms: Double = 7500
        /// Caps how hard a weak signal — and its noise — gets pushed.
        var maxGain: Double = 16
        /// Lets a hot mode be attenuated rather than only boosted.
        var minGain: Double = 0.05
        /// Falling gain (too loud) is fast; rising gain (too quiet) is slow.
        var attackTc: Double = 0.12
        var releaseTc: Double = 1.5
        /// Below this input RMS the gain is held, so silence and noise do not
        /// get pumped up.
        var noiseFloorRms: Double = 40
        var enabled = false
    }

    private(set) var gain: Double = 1
    var config = Config()

    /// RMS over the makeup-applied buffer, interleaved samples included.
    static func rms(_ pcm: [Float], makeup: Double) -> Double {
        guard !pcm.isEmpty else { return 0 }
        var sumSq: Double = 0
        for v in pcm {
            let s = Double(v) * makeup
            sumSq += s * s
        }
        return (sumSq / Double(pcm.count)).squareRoot()
    }

    private func desiredGain(_ rms: Double) -> Double? {
        guard rms >= config.noiseFloorRms else { return nil }
        return min(max(config.targetRms / rms, config.minGain), config.maxGain)
    }

    /// Advances the gain one buffer toward the target. `dt` is the buffer
    /// duration in seconds.
    @discardableResult
    func observe(_ pcm: [Float], makeup: Double, dt: Double) -> Double {
        guard config.enabled else { gain = 1; return gain }
        guard let desired = desiredGain(Self.rms(pcm, makeup: makeup)) else { return gain }
        let tc = desired < gain ? config.attackTc : config.releaseTc
        let alpha = dt > 0 && tc > 0 ? 1 - exp(-dt / tc) : 1
        gain += (desired - gain) * alpha
        return gain
    }

    /// Jumps straight to the target with no time constant — for unmute, so
    /// audio returns at level instead of crawling up the slow release.
    @discardableResult
    func snap(_ pcm: [Float], makeup: Double) -> Double {
        guard config.enabled else { gain = 1; return gain }
        guard let desired = desiredGain(Self.rms(pcm, makeup: makeup)) else { return gain }
        gain = desired
        return gain
    }

    func reset() { gain = 1 }
}
